### PR TITLE
JavaScript: Use API graphs for library modelling

### PR DESCRIPTION
**Pull request by max-schaefer**
This PR contributes an implementation of API graphs, which are a way of describing the API surface produced and/or consumed by a code base. The nodes of the API graph, referred to as "API features", represent uses and definitions of API components like functions exported by npm packages or their parameters and return values. Edges are directed and labelled, and indicate how the features relate to each other.

As a concrete example, consider this code snippet:

```js
const fs = require('fs');
fs.readFile('data.csv', (err, data) => {
  if (err) throw err;
  console.log(data);
});
```

It is described by this API graph (very slightly simplified):

api-graph

Nodes rendered as boxes represent definitions, diamonds are uses; the oval root node is neither a use nor a definition. I have labelled nodes with the code snippet they correspond to where possible (the node representing the use of module `fs` and the root node do not have corresponding code).

Note that by reading off the labels on a path from the root to a node you get a global access path for the component used/defined by this node: for example, the `err` parameter is the 0th parameter of the (callback passed as) 1st parameter to member `readFile` of the exports of module `fs`, so its access path is `/module fs/member exports/member readFile/parameter 1/parameter 0` (your notation may vary).

When the API graph is a tree, as in this case, it is just a representation of a (finite) set of access paths. In general, however, API graphs can be DAGs or even completely general graphs with cycles. In the former case, a single node can have multiple access paths, and in the latter it can even have infinitely many access paths. Put another way, API graphs represent a possibly infinite set of access paths, and encode aliasing relationship between them.

One practical application of API graphs is in library modelling, where they can be used as a DSL to describe how a code base accesses a library API. This PR contributes a library for doing that, which would allow us, for example, to abstractly describe the parameter `err` in the example above as

```ql
API::moduleImport("fs").getMember("readFile").getParameter(1).getParameter(0)
```

Note that the sequence of QL predicate calls essentially just encodes the access path, with `moduleImport` conveniently compressing the first two steps into one.

Once we have selected an API feature in this way, we can then use the predicates `getAUse()` and `getARhs()` to map it to a data-flow node representing either a use or (the right-hand side of) a definition of the corresponding API feature.

The present PR shows how to port our libraries for modelling SQL connectors and command-execution libraries to make use of API graphs.

The look-and-feel is very similar to our current `SourceNode`-based approach, but while that approach only performs local data flow and inter-procedural flow has to be added manually via type tracking, API graphs come with inter-procedural flow built in. This scales because they only type track nodes that are reachable from the API surface, which in practice tends to work out. (I am running a final evaluation to confirm, but so far experiments show only a mild performance penalty, which will pay for itself as we port more of the library to use API graphs.)

As a concrete example where this additional power is useful, this PR also switches `MissingRateLimiting.qll` to API graphs, fixing #4000 without the syntactic overhead additional type tracking would require.

### CODE REVIEW

**Comment by hvitved** [comment]
> This PR contributes an implementation of API graphs, which are a way of describing the API surface of a library both in terms of its implementation and in terms of its uses. Technically, it basically amounts to an implementation of global access paths enriched with aliasing information. (More complete description TBD.)

I wonder how this relates to [how we describe flow summaries in C#](https://github.com/github/codeql/blob/main/csharp/ql/src/semmle/code/csharp/dataflow/LibraryTypeDataFlow.qll#L338-L351). We also recently added access path sensitivity to flow summaries.

**Comment by max-schaefer** [comment]
I don't think it's very closely related (though I'm basing this on an admittedly somewhat superficial understanding of the C# flow summaries). The goal of this PR is to identify data-flow nodes that are reads/writes of global access paths, whereas the flow summaries deal, I think, primarily with local access paths.

For example, we want to be able to describe the call-back parameter `conn` in the following snippet:

```js
require('mysql').createPool().getConnection(function(err, conn) { ... })
```

And crucially we want this to work even if the calls aren't nicely chained as in this example, but local or even global data flow is needed to connect up the individual steps.

Global access paths can also be used to describe flow summaries, but that's not in scope for this PR.

(Apologies for the under-documented state of this; I'll add more details as I work towards moving it out of draft status.)

File: javascript/ql/src/ApiGraphs.qll [hunk 1]
```diff
@@ -0,0 +1,781 @@
...
+          or
+          lbl = Label::promised() and
+          rhs = pred.(PromiseCreationCall).getValue() and
+          not pred instanceof PromiseAllCreation
+          or
+          lbl = Label::promised() and
+          exists(DataFlow::MethodCallNode m | m = pred |
+            m.getMethodName() = "then" and
+            rhs = m.getCallback([0 .. 1]).getAReturn()
+          )
```
[the remark this answers is not in the document]
Comment (max-schaefer) [reply javascript/ql/src/ApiGraphs.qll:411]: Ah, nice, thanks!

**Comment by max-schaefer** [comment]
Just in time for the weekend, I have some encouraging [early performance numbers](https://gist.github.com/max-schaefer/e35bf6d38cc1788ae7bfc2f4ad0b03ca), which suggest a mild overall slowdown. Getting to neutral performance would be great, but I think for now the important thing is that it doesn't completely blow up. (Also note that relative performance will improve as more individual type trackers are replaced by API graphs.)

File: javascript/ql/src/ApiGraphs.qll [hunk 2]
```diff
@@ -0,0 +1,749 @@
...
+        succ = MkCanonicalNameUse(cn) or
+        succ = MkCanonicalNameDef(cn)
+      )
+      or
+      exists(CanonicalName cn1, CanonicalName cn2 |
+        cn2 = cn1.getAChild() and
+        lbl = Label::member(cn2.getName())
+      |
+        (pred = MkCanonicalNameDef(cn1) or pred = MkCanonicalNameUse(cn1)) and
+        (succ = MkCanonicalNameDef(cn2) or succ = MkCanonicalNameUse(cn2))
```
Comment (asgerf) [inline javascript/ql/src/ApiGraphs.qll:653]: Perhaps set literals would make this a little nicer

File: javascript/ql/src/semmle/javascript/frameworks/SQL.qll [hunk 3]
```diff
@@ -87,8 +80,9 @@ private module MySql {
     string kind;
 
     Credentials() {
-      exists(string prop |
-        this = [createConnection(), createPool()].getOptionArgument(0, prop).asExpr() and
+      exists(API::Feature call, string prop |
+        (call = createConnection() or call = createPool()) and
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/frameworks/SQL.qll:84]: Is there some reason to go from set literal to `(... or ...)` here?

File: javascript/ql/src/semmle/javascript/frameworks/SystemCommandExecutors.qll [hunk 4]
```diff
@@ -51,13 +56,19 @@ private class SystemCommandExecutors extends SystemCommandExecution, DataFlow::I
...
         shell = true and
-        mod = "exec" and
-        optionsArg = -2 and
-        cmdArg = 0
+        (
+          mod = "exec" and
+          optionsArg = -2 and
+          cmdArg = 0
+          or
+          mod = "remote-exec" and
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/frameworks/SystemCommandExecutors.qll]: Did you mean to sneak in support for `remote-exec` here? Which is fine of course, just checking.

File: javascript/ql/src/semmle/javascript/security/dataflow/MissingRateLimiting.qll [hunk 5]
```diff
@@ -148,9 +145,9 @@ class BruteForceRateLimit extends RateLimiter {
  */
 class RouteHandlerLimitedByExpressLimiter extends RateLimitedRouteHandlerExpr {
   RouteHandlerLimitedByExpressLimiter() {
-    exists(DataFlow::ModuleImportNode expressLimiter |
-      expressLimiter.getPath() = "express-limiter" and
-      expressLimiter.getACall().getArgument(0).getALocalSource().asExpr() =
+    exists(API::Feature expressLimiter |
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/security/dataflow/MissingRateLimiting.qll:148]: It seems we don't need this `exists` anymore

**Review by asgerf — COMMENTED** [review]
This looks great! Pretty much LGTM, just some superficial bits to consider.

File: javascript/ql/src/ApiGraphs.qll (hunk 2 above)
Comment (max-schaefer) [reply javascript/ql/src/ApiGraphs.qll:653]: Doesn't immediately work due to type inference failure. Do you want me to define auxiliary predicates wrapping `MkCanonicalName{Def,Use}` so I can use them here?

File: javascript/ql/src/semmle/javascript/frameworks/SQL.qll (hunk 3 above)
Comment (max-schaefer) [reply javascript/ql/src/semmle/javascript/frameworks/SQL.qll:84]: No; set literal would be much better.

File: javascript/ql/src/semmle/javascript/frameworks/SystemCommandExecutors.qll (hunk 4 above)
Comment (max-schaefer) [reply javascript/ql/src/semmle/javascript/frameworks/SystemCommandExecutors.qll]: Apparently I managed to duplicate that (it's already defined further down). Will fix.

File: javascript/ql/src/ApiGraphs.qll (hunk 2 above)
Comment (asgerf) [reply javascript/ql/src/ApiGraphs.qll:668]: Ah of course, let's just leave it like this then.

**Comment by asgerf** [comment]
Thanks for the updates, LGTM, I'd be happy to merge now and iterate further as we migrate library models to use API graphs.

File: javascript/ql/src/ApiGraphs.qll [hunk 6]
```diff
@@ -0,0 +1,764 @@
...
+        result = pkg.getMainModule() and
+        not result.isExterns() and
+        m = pkg.getPackageName()
+      )
+    }
+
+    private predicate isUsed(CanonicalName n) {
+      exists(n.(TypeName).getAnAccess()) or
+      exists(n.(Namespace).getAnAccess()) or
+      exists(InvokeExpr invk | ast_node_symbol(invk, n))
```
Comment (erik-krogh) [inline javascript/ql/src/ApiGraphs.qll:375]:
```suggestion
      exists(InvokeExpr invk | invk.getResolvedCalleeName() = n)
```

File: javascript/ql/src/semmle/javascript/frameworks/SQL.qll [hunk 7]
```diff
@@ -230,30 +197,24 @@ private module Sqlite {
...
-  /** Gets a data flow node referring to a request object. */
-  DataFlow::SourceNode request() { result = request(DataFlow::TypeTracker::end()) }
-
   /** A tagged template evaluated as a query. */
   private class QueryTemplateExpr extends DatabaseAccess, DataFlow::ValueNode {
     override TaggedTemplateExpr astNode;
 
-    QueryTemplateExpr() { mssql().getAPropertyRead("query").flowsToExpr(astNode.getTag()) }
+    QueryTemplateExpr() {
+      mssql().getMember("query").getAUse().(DataFlow::SourceNode).flowsToExpr(astNode.getTag())
```
Comment (erik-krogh) [inline javascript/ql/src/semmle/javascript/frameworks/SQL.qll:216]:
```suggestion
      mssql().getMember("query").getAUse() = DataFlow::valueNode(astNode.getTag())
```

**Comment by max-schaefer** [comment]
I have a few minor cleanups I'd like to add, a fix for a regression caused by making `module` and `exports` source nodes, and a bunch of unit tests. Then I'll do another performance evaluation over the weekend just to be sure, and if everything goes well perhaps we can merge on Monday?

**Comment by max-schaefer** [comment]
I've added the cleanups and the tests, in self-contained commits to ease reviewing.

Apart from the outstanding evaluation, there is one issue I'd like to see resolved before this go ahead, which is the inevitable terminology bikeshed: what do people think about the "API feature" terminology, and in particular the naming of the class `API::Feature`?

Clearly, API features are just nodes of the API graph, but my thinking was that we already have tons of different things called "nodes" (AST nodes, CFG nodes, data-flow graph nodes, and probably a few I've missed), and I didn't want to add to that pile. However, I find myself slipping back into "node" terminology quite frequently.

Thoughts? Should we stick with `API::Feature`, or just give up and rename it to `API::Node`?

**Comment by max-schaefer** [comment]
Slightly awkward restructuring needed to make the tests pass on both the old and the new test runner (cf last commit)...

**Comment by asgerf** [comment]
>  Should we stick with API::Feature, or just give up and rename it to API::Node?

Actually, yes I think we should rename it. We already use _graph_ and _edge_ terminology, and I thought some of the doc comments ended up sounding a little weird due to the word "feature". We have many things called node but at least we can be consistent about it.

**Comment by max-schaefer** [comment]
OK, I've renamed accordingly and updated a few out-of-date comments while I was at it.

**Review by asgerf — APPROVED** [review]
LGTM

**Comment by max-schaefer** [comment]
> I'll do another performance evaluation over the weekend just to be sure, and if everything goes well (...)

The stars haven't quite aligned yet; investigating a 20% slowdown on TypeScript... (Everything else looks fine, fortunately.)

**Comment by max-schaefer** [comment]
@asgerf: Performance is looking happy again (see link in internal issue). Would you like to take another look at the recent commits?

**Review by asgerf — APPROVED** [review]
LGTM. Let's merge!

**Pull request merged**